File: aoc/2020/aoc202002.hs
import Data.Bits(xor)
import Data.Char(isDigit)

parse :: String -> [(Int,Int,Char,String)]
parse = map (p . words . map (\ c -> if c == '-' then ' ' else c)) . lines
  where
    p (low:high:char:pwd:_) = (read low,read high,head char,pwd)
    p line = error (show line)

valid :: (Int,Int,Char,String) -> Bool
valid (low,high,char,pwd) = count >= low && count <= high
  where count = length (filter (== char) pwd)

testData :: String
testData = "1-3 a: abcde\n1-3 b: cdefg\n2-9 c: ccccccccc\n"

test :: ()
test
  | (length . filter valid . parse) testData /= 2 = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap (length . filter valid . parse) $ readFile "input/02.txt"

valid2 :: (Int,Int,Char,String) -> Bool
valid2 (i,j,char,pwd) = (char == pwd !! (i-1)) `xor` (char == pwd !! (j-1))

test2 :: ()
test2
  | (length . filter valid2 . parse) testData /= 1 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap (length . filter valid2 . parse) $ readFile "input/02.txt"

File: aoc/2025/aoc202002.go
package main

func init() {
	Register(&aoc202002{
		AOC: AOC{
			Day:           2,
			InputFilename: "../2020/input/02.txt",
			Tests: []Test{
				Test{
					Input: `1-3 a: abcde
1-3 b: cdefg
2-9 c: ccccccccc
`,
					Part1: "2",
					Part2: "1",
				},
			},
		},
	})
}

type aoc202002 struct {
	AOC
}

func (aoc *aoc202002) Part1(input *Input) string {
	result := 0
	for {
		n1, ok := input.Int()
		if !ok {
			break
		}
		n2, _ := input.Int()
		n2 = -n2
		w, _ := input.Word()
		pw, _ := input.Word()
		n := 0
		for i := range pw {
			if pw[i] == w[0] {
				n++
			}
		}
		if n >= n1 && n <= n2 {
			result++
		}
	}
	return IntResult(result)
}

func (aoc *aoc202002) Part2(input *Input) string {
	result := 0
	for {
		n1, ok := input.Int()
		if !ok {
			break
		}
		n2, _ := input.Int()
		n2 = -n2
		w, _ := input.Word()
		pw, _ := input.Word()
		if (pw[n1-1] == w[0]) != (pw[n2-1] == w[0]) {
			result++
		}
	}
	return IntResult(result)
}
